### ISSUE TITLE
Consider making remove unused members analyzers conservative

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#30996]
See the thread here: https://github.com/dotnet/roslyn/pull/30982#discussion_r231239939

The proposal is to make the analyzer conservative, so it assumes all private symbols are _fine_ to be unused by default, except the one we know are know by iterating them. There are pros and cons of conservative versus aggressive approach that are listed there.

**Comment by CyrusNajmabadi** [comment]
I'm fairly shocked and dismayed this even needs discussion.  This analyzer/fixer removes user code.  It is, by it's very nature, a destructive operation.  We should be conservative and should only be removing things we feel we have proven safe to remove.  INstead, it's the opposite.  If we haven't demonstrated why we should not be removing it, the language constructed is defaulted as safe to remove.

This violates all appropriate practices around analyzers/fixers, and it's super unclear to me why we need to have this discussion.  If this was a much more minor feature, this might be ok.  But this literally goes and deletes members whole-hog if invoked (i.e. with fix-all).  The bar is super high on that.  

Also, the change being requested here is extremely minimal and non-onerous.  I'm surprised htis is contentious as it helps makes things clearer, simpler and safer.

**Comment by mavasani** [comment]
The primary argument is there is no unit test being proposed whose behavior will change with your proposed approach. Can you please create a unit test that fails with the current approach (except the known ones)?

**Comment by CyrusNajmabadi** [comment]
That's not a sufficient argument to me.  The claim isn't that this has bugs.  The claim is that this is much harder to validate and maintain**, and that it's much *riskier* because it adopts a strategy where mistakes lead to customer code-loss, versus customers just not seeing the feature offered.

** This is not a hypothetical either.  The current approach forced me to have to go and examine all these different cases to try to to ensure there wasn't something missed.

It also follows the wrong approach ebcause if we *add* anything new in the future, then we may delete user code unless sufficient testing happens to ensure we update this block.  As opposed to taking an approach where if we miss a new construct, we fall back to a safer state.  This is literally a feature which offers to remove code.  I have no idea why we would ever design it ot potentially produce more false-positives than false-negatives.  It simply goes against any sort of rational practice for this sort of thing.

**Comment by mavasani** [comment]
I have switched the analyzer to be more conservative to unblock the PR #30982. I still believe a new member symbol kind or a new method kind should be an unused member candidate by default if it is private, and should only be excluded by explicitly changing the analyzer and adding unit tests case for exception cases where we know it is fine to have such a private symbol without references. I will not pollute this PR further and leave that discussion in https://github.com/dotnet/roslyn/issues/30996.

**Comment by mavasani** [comment]
This is no longer needed, I changed the original PR to make the analyzer conservative.
